fri/src/verifier.rs

```diff
@@ -0,0 +1,282 @@
+use randomness::Prng;
+use sha3::Digest;
+
+use crate::{lde::MultiplicativeLDE, parameters::FriParameters};
+use channel::{fs_verifier_channel::FSVerifierChannel, Channel, VerifierChannel};
+use commitment_scheme::{
+    make_commitment_scheme_verifier, table_verifier::TableVerifier, CommitmentHashes,
+};
+
+#[allow(dead_code)]
+pub type FirstLayerQueriesCallback<F> = fn(&[u64]) -> Vec<F>;
+
+#[allow(dead_code)]
+pub struct FriVerifier<
+    F: FftField + PrimeField,
+    P: Prng + Clone + 'static,
+    W: Digest + Clone + 'static,
+> {
+    channel: FSVerifierChannel<F, P, W>,
+    params: FriParameters<F, Radix2EvaluationDomain<F>>,
+    commitment_hashes: CommitmentHashes,
+    first_layer_callback: FirstLayerQueriesCallback<F>,
+    n_layers: usize,
+    first_eval_point: Option<F>,
+    eval_points: Vec<F>,
+    table_verifiers: Vec<TableVerifier<F, P, W>>,
+    query_indices: Vec<u64>,
+    query_results: Vec<F>,
+    expected_last_layer: Option<Vec<F>>,
+    eval_points_test: Option<Vec<F>>,
+    last_layer_coefficients_test: Option<Vec<F>>,
+}
+
+#[allow(dead_code)]
+impl<F: FftField + PrimeField, P: Prng + Clone + 'static, W: Digest + Clone + 'static>
+    FriVerifier<F, P, W>
+{
+    pub fn new(
+        channel: FSVerifierChannel<F, P, W>,
+        params: FriParameters<F, Radix2EvaluationDomain<F>>,
+        commitment_hashes: CommitmentHashes,
+        first_layer_callback: FirstLayerQueriesCallback<F>,
+        eval_points_test: Option<Vec<F>>,
+        last_layer_coefficients_test: Option<Vec<F>>,
+    ) -> Self {
+        let n_layers = params.fri_step_list.len();
+        Self {
+            channel,
+            params,
+            commitment_hashes,
+            first_layer_callback,
+            n_layers,
+            first_eval_point: None,
+            eval_points: vec![F::zero(); n_layers - 1],
+            table_verifiers: vec![],
+            query_indices: vec![],
+            query_results: vec![],
+            expected_last_layer: None,
+            eval_points_test,
+            last_layer_coefficients_test,
+        }
+    }
+
+    pub fn verify_fri(&mut self) -> Result<(), Box<dyn Error>> {
+        self.commitment_phase()?;
+
+        // // query phase
+        //self.query_indices = self.choose_query_indices();
+        // //self.channel.begin_query_phase();
+
+        // // decommitment phase
+        // // TODO : annotation
+
+        // self.verify_first_layer();
+        // self.verify_inner_layers();
+        // self.verify_last_layer();
+        Ok(())
+    }
+
+    pub fn commitment_phase(&mut self) -> Result<(), Box<dyn Error>> {
+        let mut basis_index = 0;
+        for i in 0..self.n_layers {
+            let cur_fri_step = self.params.fri_step_list[i];
+            basis_index += cur_fri_step;
+
+            if i == 0 {
+                if self.params.fri_step_list[0] != 0 {
+                    self.first_eval_point = Some(self.channel.draw_felem());
```

Comment:
What's special about the first one?

Reply:
Nothing, but will change this in next PR